### PR TITLE
NVA channel registry search

### PR DESCRIPTION
**Pull request by janbj**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,6 @@
+.aws-sam
+.idea
+*.iml
+target
+.gradle
+build
```
Comment (brinxmat) [inline .gitignore]: Typically, we use e.g. https://gitignore.io/api/java,linux,macos,gradle,windows,eclipse,intellij+all,visualstudiocode

File: FindChannelFunction/src/main/java/no/unit/nva/channel/ChannelRegistryClient.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+        try (CloseableHttpResponse response = httpClient.execute(request)) {
+            HttpEntity entity = response.getEntity();
+            if (entity == null) {
+                throw new NoResultsFoundException("Response from registry is empty.");
+            }
+            String entityString = EntityUtils.toString(entity);
+            JsonNode json = objectMapper.readTree(entityString);
+            validateJsonResponse(json);
+            json.forEach(result -> {
```
Comment (brinxmat) [inline FindChannelFunction/src/main/java/no/unit/nva/channel/ChannelRegistryClient.java]: Expression lambda: 
``` json.forEach(result -> results.add(toChannel(result)));```

File: FindChannelFunction/src/main/java/no/unit/nva/channel/ChannelRegistryClient.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    public List<Channel> fetchChannels(Integer tableId, String searchTerm) throws IOException, NoResultsFoundException {
+        FetchJsonTableDataRequest fetchJsonTableDataRequest = FetchJsonTableDataRequest.create(tableId, searchTerm);
+        List<Channel> results = new ArrayList<>();
+        HttpPost request = new HttpPost(url);
+        request.setHeader(ACCEPT, APPLICATION_JSON.getMimeType());
+        request.setHeader(CONTENT_TYPE, APPLICATION_JSON.getMimeType());
+        request.setEntity(new StringEntity(objectMapper.writeValueAsString(fetchJsonTableDataRequest)));
+
+        try (CloseableHttpResponse response = httpClient.execute(request)) {
+            HttpEntity entity = response.getEntity();
```
Comment (brinxmat) [inline FindChannelFunction/src/main/java/no/unit/nva/channel/ChannelRegistryClient.java]: Something like ```HttpEntity entity = Optional.ofNullable(response.getEntity())
                    .orElseThrow(() -> new NoResultsFoundException(NO_RESULTS_FROM_SERVICE));```

Although to my mind, the naming here does not imply that the service is in error in the way I believe it is (this is a system error that should result in a 502 status code? Unless I have misunderstood).

File: FindChannelFunction/src/main/java/no/unit/nva/channel/ChannelRegistryClient.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            validateJsonResponse(json);
+            json.forEach(result -> {
+                results.add(toChannel(result));
+            });
+        }
+        return results;
+    }
+
+    private void validateJsonResponse(JsonNode jsonResponse) throws NoResultsFoundException {
+        if (jsonResponse.has(1)
```
Comment (brinxmat) [inline FindChannelFunction/src/main/java/no/unit/nva/channel/ChannelRegistryClient.java]: If we extract all of the magics into constants with descriptive names, I will be happy, but I guess we also have feedback for the API designers here :D

File: FindChannelFunction/src/main/java/no/unit/nva/channel/ChannelRegistryClient.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+                && jsonResponse.get(1).at("/returnerte poster").isNumber()
+                && jsonResponse.get(1).at("/returnerte poster").intValue() == 0) {
+            throw new NoResultsFoundException("No results found for this search term.");
+        }
+    }
+
+    protected Channel toChannel(JsonNode json) {
```
Comment (brinxmat) [inline FindChannelFunction/src/main/java/no/unit/nva/channel/ChannelRegistryClient.java:88]: By and large, this is an argument for mapping to objects in the gson style in my mind, but extract constants

File: FindChannelFunction/src/main/java/no/unit/nva/channel/Environment.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package no.unit.nva.channel;
+
+import java.util.Optional;
+
+public class Environment {
+
+    public Optional<String> get(String name) {
```
Comment (brinxmat) [inline FindChannelFunction/src/main/java/no/unit/nva/channel/Environment.java:13]: I think ```return Optional.ofNullable(System.getenv(name)).filter(s -> !s.isEmpty());``` covers it as Optional.ofNullable returns Optional.empty()

File: FindChannelFunction/src/main/java/no/unit/nva/channel/FindChannelFunctionApp.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        this.channelRegistryClient = channelRegistryClient;
+        this.allowedOrigin = environment.get("ALLOWED_ORIGIN").orElseThrow(IllegalStateException::new);
+    }
+
+    @Override
+    public void handleRequest(InputStream input, OutputStream output, Context context) throws IOException {
+        SearchRequest searchRequest;
+        try {
+            JsonNode event = objectMapper.readTree(input);
+            searchRequest = objectMapper.readValue(event.get("body").asText(), SearchRequest.class);
```
Comment (brinxmat) [inline FindChannelFunction/src/main/java/no/unit/nva/channel/FindChannelFunctionApp.java]: What is "body" here?

File: FindChannelFunction/src/test/java/no/unit/nva/channel/EnvironmentTest.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+import org.junit.Test;
+import org.junit.contrib.java.lang.system.EnvironmentVariables;
+
+import java.util.Optional;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+
+public class EnvironmentTest {
+
```
Comment (brinxmat) [inline FindChannelFunction/src/test/java/no/unit/nva/channel/EnvironmentTest.java:13]: Add test for empty string.

File: FindChannelFunction/src/main/java/no/unit/nva/channel/FindChannelFunctionApp.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+    public FindChannelFunctionApp() {
+        this(createObjectMapper(), new ChannelRegistryClient(createObjectMapper(), HttpClients.createDefault(),
+                CHANNEL_REGISTRY_URI), new Environment());
+    }
+
+    public FindChannelFunctionApp(ObjectMapper objectMapper, ChannelRegistryClient channelRegistryClient, Environment environment) {
+        this.objectMapper = objectMapper;
+        this.channelRegistryClient = channelRegistryClient;
+        this.allowedOrigin = environment.get("ALLOWED_ORIGIN").orElseThrow(IllegalStateException::new);
```
Comment (brinxmat) [inline FindChannelFunction/src/main/java/no/unit/nva/channel/FindChannelFunctionApp.java]: Read ALLOWED_ORIGIN as a constant at first pass, so probably should be one

File: FindChannelFunction/src/main/java/no/unit/nva/channel/FindChannelFunctionApp.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    @Override
+    public void handleRequest(InputStream input, OutputStream output, Context context) throws IOException {
```
Comment (brinxmat) [inline FindChannelFunction/src/main/java/no/unit/nva/channel/FindChannelFunctionApp.java]: Do we ever check that context is not null? It shouldn't be, but my IDE tells me it might be

File: FindChannelFunction/src/main/java/no/unit/nva/channel/FindChannelFunctionApp.java [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+        try {
+            List<Channel> channels = channelRegistryClient.fetchChannels(searchRequest.getTableId(),
+                    searchRequest.getSearchTerm());
+            SearchResponse searchResponse = new SearchResponse(channels);
+            writeResponse(output, searchResponse, SC_OK);
+        } catch (NoResultsFoundException e) {
+            writeErrorResponse(output, Problem.valueOf(NOT_FOUND, e.getMessage(), createProblemInstance(context)),
+                    SC_NOT_FOUND);
+        } catch (Exception e) {
+            System.out.println(e);
```
Comment (brinxmat) [inline FindChannelFunction/src/main/java/no/unit/nva/channel/FindChannelFunctionApp.java]: We get a warning here about sending Throwable e to System.out, which I guess we need to suppress in our config for Lambdas

File: FindChannelFunction/src/main/java/no/unit/nva/channel/FindChannelFunctionApp.java [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+            writeErrorResponse(output, Problem.valueOf(INTERNAL_SERVER_ERROR, e.getMessage(),
+                    createProblemInstance(context)), SC_INTERNAL_SERVER_ERROR);
+        }
+    }
+
+    private void writeResponse(OutputStream output, Object body, int statusCode) throws IOException {
```
Comment (brinxmat) [inline FindChannelFunction/src/main/java/no/unit/nva/channel/FindChannelFunctionApp.java]: At the present time, statusCode is always 200, however, do we see any changes in this? I doubt it, but worth checking

File: FindChannelFunction/src/main/java/no/unit/nva/channel/FindChannelFunctionApp.java [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+        Map<String,String> headers = new ConcurrentHashMap<>();
```
Comment (brinxmat) [inline FindChannelFunction/src/main/java/no/unit/nva/channel/FindChannelFunctionApp.java]: writeResponse and writeErrorResponse can be a single method filtering statusCode to provide appropriate content-type

File: FindChannelFunction/src/main/java/no/unit/nva/channel/FindChannelFunctionApp.java [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+            return new URIBuilder()
+                    .setHost(context.getAwsRequestId())
+                    .setPath(UUID.randomUUID().toString())
+                    .build();
+        } catch (URISyntaxException e) {
+            System.out.println(e);
+            return null;
+        }
+    }
+}
```
Comment (brinxmat) [inline FindChannelFunction/src/main/java/no/unit/nva/channel/FindChannelFunctionApp.java]: Newline :)

File: FindChannelFunction/src/main/java/no/unit/nva/channel/FindChannelFunctionApp.java [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+        return new ObjectMapper().registerModule(new ProblemModule());
+    }
+
+    public static URI createProblemInstance(Context context) {
+        try {
+            return new URIBuilder()
+                    .setHost(context.getAwsRequestId())
+                    .setPath(UUID.randomUUID().toString())
+                    .build();
+        } catch (URISyntaxException e) {
```
Comment (brinxmat) [inline FindChannelFunction/src/main/java/no/unit/nva/channel/FindChannelFunctionApp.java]: Can you add a comment as to why we are not handling the error here?

File: FindChannelFunction/src/main/java/no/unit/nva/channel/FindChannelFunctionApp.java [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+            writeErrorResponse(output, Problem.valueOf(INTERNAL_SERVER_ERROR, e.getMessage(),
+                    createProblemInstance(context)), SC_INTERNAL_SERVER_ERROR);
+        }
+    }
+
+    private void writeResponse(OutputStream output, Object body, int statusCode) throws IOException {
+        Map<String,String> headers = new ConcurrentHashMap<>();
+        headers.put(CONTENT_TYPE, APPLICATION_JSON.getMimeType());
+        headers.put(ACCESS_CONTROL_ALLOW_ORIGIN, allowedOrigin);
+        headers.put(HttpHeaders.VARY, "Origin");
```
Comment (brinxmat) [inline FindChannelFunction/src/main/java/no/unit/nva/channel/FindChannelFunctionApp.java]: "Vary: Origin" was completely unknown to me, I'd still like Origin as a Constant

File: FindChannelFunction/src/main/java/no/unit/nva/channel/exception/NoResultsFoundException.java [hunk 17]
```diff
@@ -0,0 +1,8 @@
+package no.unit.nva.channel.exception;
+
+public class NoResultsFoundException extends Exception {
```
Comment (brinxmat) [inline FindChannelFunction/src/main/java/no/unit/nva/channel/exception/NoResultsFoundException.java:3]: I got the impression that we might be using this for two distinct classes of error

File: FindChannelFunction/src/main/java/no/unit/nva/channel/model/incoming/SearchRequest.java [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class SearchRequest {
+
+    private final Integer tableId;
+    private final String searchTerm;
+
+    public SearchRequest(
+            @JsonProperty(value = "tableId", required = true) Integer tableId,
```
Comment (brinxmat) [inline FindChannelFunction/src/main/java/no/unit/nva/channel/model/incoming/SearchRequest.java:13]: The naming tableId is indicative of something in the resource we're addressing…

File: FindChannelFunction/src/main/java/no/unit/nva/channel/model/outgoing/Channel.java [hunk 19]
```diff
@@ -0,0 +1,62 @@
+package no.unit.nva.channel.model.outgoing;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class Channel {
+
+    @JsonProperty("title")
+    private String originalTitle;
```
Comment (brinxmat) [inline FindChannelFunction/src/main/java/no/unit/nva/channel/model/outgoing/Channel.java:8]: What are you implying with "originalTitle"?

File: FindChannelFunction/src/main/java/no/unit/nva/channel/model/outgoing/Channel.java [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+
+    @JsonProperty("onlineIssn")
+    private String onlineIssn;
```
Comment (brinxmat) [inline FindChannelFunction/src/main/java/no/unit/nva/channel/model/outgoing/Channel.java:11]: e-issn is a relatively standardised concept, and the other is issn. I suspect that spelling this out in this way is fair enough, there is no ambiguity

File: FindChannelFunction/src/main/java/no/unit/nva/channel/model/outgoing/SearchResponse.java [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+    }
+
+    public List<Channel> getResults() {
+        return results;
+    }
+
+    public void setResults(List<Channel> results) {
+        this.results = results;
+    }
+}
```
Comment (brinxmat) [inline FindChannelFunction/src/main/java/no/unit/nva/channel/model/outgoing/SearchResponse.java]: Newline :)

File: FindChannelFunction/src/test/java/no/unit/nva/channel/ChannelRegistryClientTest.java [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class ChannelRegistryClientTest {
+
+    private ObjectMapper objectMapper = FindChannelFunctionApp.createObjectMapper();
+
+    @Test(expected = NoResultsFoundException.class)
```
Comment (brinxmat) [inline FindChannelFunction/src/test/java/no/unit/nva/channel/ChannelRegistryClientTest.java]: I started using Jupiter in another project, this allows better methods for testing exceptions

File: FindChannelFunction/src/main/java/no/unit/nva/channel/GatewayResponse.java [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+    private Map<String, String> headers;
+    private int statusCode;
+
+    public GatewayResponse() {
+    }
+
+    public GatewayResponse(final T body, final Map<String, String> headers, final int statusCode) {
+        this.statusCode = statusCode;
+        this.body = body;
+        this.headers = Collections.unmodifiableMap(new HashMap<>(headers));
```
Comment (brinxmat) [inline FindChannelFunction/src/main/java/no/unit/nva/channel/GatewayResponse.java:30]: Map::copyOf, apparently

File: FindChannelFunction/src/main/java/no/unit/nva/channel/GatewayResponse.java [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+public class GatewayResponse<T> {
+
+    private T body;
+    private Map<String, String> headers;
+    private int statusCode;
+
+    public GatewayResponse() {
```
Comment (brinxmat) [inline FindChannelFunction/src/main/java/no/unit/nva/channel/GatewayResponse.java]: Unused constructors and methods, but I guess you have your reasons

File: FindChannelFunction/src/main/java/no/unit/nva/channel/ChannelRegistryClient.java (hunk 3 above)
Comment (janbj) [reply FindChannelFunction/src/main/java/no/unit/nva/channel/ChannelRegistryClient.java]: Dersom tjenesten hadde svart med en feil så er jeg enig, men her er det bare en respons uten treff.

File: FindChannelFunction/src/main/java/no/unit/nva/channel/ChannelRegistryClient.java (hunk 5 above)
Comment (janbj) [reply FindChannelFunction/src/main/java/no/unit/nva/channel/ChannelRegistryClient.java:88]: Fjernet magic strings, men avventer tilbakemelding fra API designerne for resten.

File: FindChannelFunction/src/main/java/no/unit/nva/channel/ChannelRegistryClient.java (hunk 4 above)
Comment (janbj) [reply FindChannelFunction/src/main/java/no/unit/nva/channel/ChannelRegistryClient.java]: Fjernet magic strings, men avventer tilbakemelding fra API designerne for resten.

File: FindChannelFunction/src/main/java/no/unit/nva/channel/model/outgoing/Channel.java (hunk 20 above)
Comment (janbj) [reply FindChannelFunction/src/main/java/no/unit/nva/channel/model/outgoing/Channel.java:11]: Speiler bare hva kanalregisteret kaller det. Vi får eventuelt ta det opp med produkteierne om vi ønsker andre navn her.

File: FindChannelFunction/src/main/java/no/unit/nva/channel/GatewayResponse.java (hunk 24 above)
Comment (janbj) [reply FindChannelFunction/src/main/java/no/unit/nva/channel/GatewayResponse.java]: Jeg har fjernet en del default constructors og setters og erstattet med JsonCreator, men getterne brukes fortsatt av jackson ved mapping.

File: FindChannelFunction/src/main/java/no/unit/nva/channel/GatewayResponse.java (hunk 23 above)
Comment (janbj) [reply FindChannelFunction/src/main/java/no/unit/nva/channel/GatewayResponse.java:30]: Ser ut som vi trenger Java 10 eller Guava.
